### ISSUE TITLE
Puc_v4_Factory::buildUpdateChecker run in front end

### ISSUE THREAD

**Issue by gaupoit** [YahnisElsts/plugin-update-checker#341]
Thanks for creating this wonderful library. We love it ❤️ 
I have a concern that if we put this snippet 

```
require 'path/to/plugin-update-checker/plugin-update-checker.php';

$myUpdateChecker = Puc_v4_Factory::buildUpdateChecker(
	'http://example.com/path/to/details.json',
	__FILE__, //Full path to the main plugin file or functions.php.
	'unique-plugin-or-theme-slug'
);
```
in the main plugin file. In every page load, it will call buildUpdateChecker in front end or admin site. Are there any performance issues with this? Should I call it in admin_init hook only?

**Comment by YahnisElsts** [comment]
I haven't measured the performance impact of running PUC in the front end, but I expect it to be relatively small. When not actively checking for updates, all it does is set a bunch of hooks and maybe read one option from the database.

Also, there are a couple of reasons why it can be useful to *always* call `buildUpdateChecker`:

- Logged-in admins that are on the site front end can see the WordPress Toolbar/Admin Bar which displays the number of available updates. If you don't initialise PUC then this number won't include your update(s).
- Site management tools like WP-CLI, ManageWP, InfiniteWP and so on may not trigger the `admin_init` hook at all. To make sure that these tools can see your updates, you'll need to either run PUC on every page load or add special case code to support each tool.
- Some regular tasks like update checks and automatic update installation happen in cron requests, so you will also need to call `buildUpdateChecker` when `DOING_CRON` is `true`.

**Comment by seshelby** [comment]
I am able to trigger it showing an update everytime I load a page so it appears to check for updates on every page load.  Is there a way to limit update checks to only twice per day?

**Comment by YahnisElsts** [comment]
Are you sure that it's actually *checking for updates* every time it shows an update? Normally, it should cache the update information and show that cached update instead of checking on every page load.

You can use the [Debug Bar](https://wordpress.org/plugins/debug-bar/) plugin to see when was the last time PUC checked for updates.

**Comment by seshelby** [comment]
I am pretty sure it is. I installed the debug plugin and it doesn't show
that a check has occurred but if I change the version number of the
installed plugin to a version less than the most resent and visit the site
the admin bar updates to show that an update is available so that would
indicate that it is doing the check wouldn't it?

Sam Shelby
AlumniOnline Web Services
www.alumnionlineservices.com

**Comment by seshelby** [comment]
Thanks for the quick response. I think I understand now. It does the check periodically and saves it somewhere which it compares the installed version to each time the page loads. I changed the version on the remote file instead and it didn't reflect the change when I reloaded the page which indicates that it is working as expected.

**Comment by YahnisElsts** [comment]
> [...] if I change the version number of the installed plugin to a version less than the most resent and visit the site the admin bar updates to show that an update is available 

In this case, changing the local version number is what makes PUC check for updates. When PUC checks for updates and stores update information, it also stores the currently installed plugin version. If it later detects that the installed version no longer matches the stored version, it will check for updates again.

The idea behind this behaviour is that seeing a different installed version is a sign that the cached update might no longer apply. For example, maybe the new installed version has different WP/PHP requirements, or different update settings, or something like that.

**Comment by seshelby** [comment]
Thank you for the quick response and clarification.

**Comment by seshelby** [comment]
Thank you for the information. It still appears to be checking periodically
when the version number has not be changed. Is there a way to disable this
behavior and only do the check when the scheduled automatic update occurs?

Sam Shelby
AlumniOnline Web Services
www.alumnionlineservices.com

**Comment by YahnisElsts** [comment]
It will also check more frequently on certain admin pages, like "Dashboard -> Updates". This mimics how WordPress itself checks for updates. Could this explain why you're seeing more frequent checks?

There are multiple ways to control how often PUC checks for updates. For example, you can set the fourth argument to `buildUpdateChecker` to zero. This will completely disable automatic checks. Then you could call the `checkForUpdates` method yourself in a custom cron job. 

Alternatively, you could use the `puc_check_now-$slug` filter and return `false` to stop a specific update request.

Finally, you could create a subclass of `Puc_v4p9_Plugin_UpdateChecker`, override the `createScheduler` method and supply your own implementation of the `Puc_v4p9_Scheduler` class that decides when to check for updates.

**Comment by seshelby** [comment]
It could be. I would like to only have it check for updates every few days
regardless of where someone is on the website. If I set the 4th parameter
to 0 and create a cron to do the automatic check will this accomplish that?

Thank you!

**Comment by YahnisElsts** [comment]
Yes, that should work. The 4th argument (i.e. the one after the slug) sets the check interval in hours. Setting it to 0 should disable the default cron job and other built-in checks.

**Comment by seshelby** [comment]
Great, Thank you!
